README.md
# osparc-python-runner-tensorflow

oSparc Python Runner Tensorflow
![Github-CI](https://github.com/ITISFoundation/osparc-python-runner-tensorflow/workflows/Github-CI%20Push/PR%20osparc-python-runner-tensorflow/badge.svg)

## Description

oSparc service that allows running any python-based script either as simple script of complex python application inside a zip file (with or without requirements.txt file)


## Development

```console
make help                                # shows Makefile help

make devenv                              # creates python virtual environment
source .venv/bin/activate                # activates python virtual environment
make build                               # builds oSparc compatible docker image
make info-build                          # shows oSparc compatible docker image
make tests                               # tests oSparc compatible docker image
```

## CI/CD Integration

### Github-Actions

[github-ci.yml](https://github.com/ITISFoundation/osparc-python-runner-tensorflow/blob/master/.github/workflows/github-ci.yml)


CHANGELOG.md
# Changelog

## [1.0.1] - 2021-07-23
### Added
 - Initial version
  

<!-- HOW TO WRITE  THIS CHANGELOG

- Guiding Principles
  - Changelogs are for humans, not machines.
  - There should be an entry for every single version.
  - The same types of changes should be grouped.
  - Versions and sections should be linkable.
  - The latest version comes first.
  - The release date of each version is displayed.
  - Mention whether you follow Semantic Versioning.
  -
- Types of changes
  - Added for new features.
  - Changed for changes in existing functionality.
  - Deprecated for soon-to-be removed features.
  - Removed for now removed features.
  - Fixed for any bug fixes.
  - Security in case of vulnerabilities.

SEE https://keepachangelog.com/en/1.0.0/
-->


tests/integration/conftest.py
# pylint:disable=unused-variable
# pylint:disable=unused-argument
# pylint:disable=redefined-outer-name

import os
from pathlib import Path

import pytest

import docker


@pytest.fixture
def docker_client() -> docker.DockerClient:
    return docker.from_env()


@pytest.fixture
def docker_image_key(docker_client: docker.DockerClient) -> str:
    image_key = "osparc-python-runner-tensorflow:"
    docker_images = [image for image in docker_client.images.list() if any(
        image_key in tag for tag in image.tags)]
    return docker_images[0].tags[0]


@pytest.fixture
def docker_image(docker_client: docker.DockerClient, docker_image_key: str) -> docker.models.images.Image:
    docker_image = docker_client.images.get(docker_image_key)
    assert docker_image
    return docker_image


def _is_gitlab_executor() -> bool:
    return "GITLAB_CI" in os.environ


@pytest.fixture
def temporary_path(tmp_path: Path) -> Path:
    if _is_gitlab_executor():
        # /builds is a path that is shared between the docker in docker container and the job builder container
        shared_path = Path(
            "/builds/{}/tmp".format(os.environ["CI_PROJECT_PATH"]))
        shared_path.mkdir(parents=True, exist_ok=True)
        return shared_path
    return tmp_path


examples/test_tensorflow_gpu.py
import tensorflow as tf
import timeit

# Test if there is a GPU availble
device_name = tf.test.gpu_device_name()
if device_name != "/device:GPU:0":
    raise SystemError("GPU device not found")

print("Found GPU at: {}".format(device_name))

# Run benchmark
device_name = tf.test.gpu_device_name()
if device_name != "/device:GPU:0":
    print(
        "\n\nThis error most likely means that this notebook is not "
        "configured to use a GPU.  Change this in Notebook Settings via the "
        "command palette (cmd/ctrl-shift-P) or the Edit menu.\n\n"
    )
    raise SystemError("GPU device not found")


def cpu():
    with tf.device("/cpu:0"):
        random_image_cpu = tf.random.normal((100, 100, 100, 3))
        net_cpu = tf.keras.layers.Conv2D(32, 7)(random_image_cpu)
        return tf.math.reduce_sum(net_cpu)


def gpu():
    with tf.device("/device:GPU:0"):
        random_image_gpu = tf.random.normal((100, 100, 100, 3))
        net_gpu = tf.keras.layers.Conv2D(32, 7)(random_image_gpu)
        return tf.math.reduce_sum(net_gpu)


# We run each op once to warm up; see: https://stackoverflow.com/a/45067900
cpu()
gpu()

# Run the op several times.
print(
    "Time (s) to convolve 32x7x7x3 filter over random 100x100x100x3 images "
    "(batch x height x width x channel). Sum of ten runs."
)
print("CPU (s):")
cpu_time = timeit.timeit("cpu()", number=10, setup="from __main__ import cpu")
print(cpu_time)
print("GPU (s):")
gpu_time = timeit.timeit("gpu()", number=10, setup="from __main__ import gpu")
print(gpu_time)
print("GPU speedup over CPU: {}x".format(int(cpu_time / gpu_time)))


tests/unit/test_validation_data.py
# pylint:disable=unused-variable
# pylint:disable=unused-argument
# pylint:disable=redefined-outer-name
import json
from pathlib import Path
from typing import Dict

import pytest
import yaml


@pytest.fixture
def port_type() -> str:
    return ""


@pytest.fixture
def label_cfg(metadata_file: Path, port_type: str) -> Dict:
    ports_type = f"{port_type}s"
    with metadata_file.open() as fp:
        cfg = yaml.safe_load(fp)
        assert ports_type in cfg
        return cfg[ports_type]


@pytest.fixture
def validation_folder(validation_dir: Path, port_type: str) -> Path:
    return validation_dir / port_type


@pytest.fixture
def validation_cfg(validation_dir: Path, port_type: str) -> Dict:
    validation_file = validation_dir / port_type / (f"{port_type}s.json")
    if validation_file.exists():
        with validation_file.open() as fp:
            return json.load(fp)
    # it may not exist if only files are required
    return None


def _find_key_in_cfg(filename: str, value: Dict) -> str:
    for k, v in value.items():
        if k == filename:
            if isinstance(v, dict):
                assert "data:" in v["type"]
                yield k
            else:
                yield v
        elif isinstance(v, dict):
            for result in _find_key_in_cfg(filename, v):
                yield result


@pytest.mark.parametrize("port_type", ["input", "output"])
def test_validation_data_follows_definition(
    label_cfg: Dict, validation_cfg: Dict, validation_folder: Path
):
    for key, value in label_cfg.items():
        assert "type" in value

        # rationale: files are on their own and other types are in inputs.json
        if not "data:" in value["type"]:
            # check that keys are available
            assert key in validation_cfg
        else:
            # it's a file and it should be in the folder as well using key as the filename
            filename_to_look_for = key
            if "fileToKeyMap" in value:
                # ...or there is a mapping
                assert len(value["fileToKeyMap"]) > 0
                for filename, mapped_value in value["fileToKeyMap"].items():
                    assert mapped_value == key
                    filename_to_look_for = filename
                    assert (validation_folder / filename_to_look_for).exists()
            else:
                assert (validation_folder / filename_to_look_for).exists()

    if validation_cfg:
        for key, value in validation_cfg.items():
            # check the key is defined in the labels
            assert key in label_cfg
            label2types = {
                "number": (float, int),
                "integer": int,
                "boolean": bool,
                "string": str,
            }
            if not "data:" in label_cfg[key]["type"]:
                # check the type is correct
                assert isinstance(value, label2types[label_cfg[key]["type"]])



tools/update_compose_labels.py
#!/bin/python

""" Update a docker-compose file with json files in a path

    Usage: python update_compose_labels --c docker-compose.yml -f folder/path

:return: error code
"""

import argparse
import json
import logging
import sys
from enum import IntEnum
from pathlib import Path
from typing import Dict

import yaml

log = logging.getLogger(__name__)
logging.basicConfig(level=logging.INFO)


class ExitCode(IntEnum):
    SUCCESS = 0
    FAIL = 1


def get_compose_file(compose_file: Path) -> Dict:
    with compose_file.open() as filep:
        return yaml.safe_load(filep)


def get_metadata_file(metadata_file: Path) -> Dict:
    with metadata_file.open() as fp:
        return yaml.safe_load(fp)


def stringify_metadata(metadata: Dict) -> Dict[str, str]:
    jsons = {}
    for key, value in metadata.items():
        jsons[f"io.simcore.{key}"] = json.dumps({key: value})
    return jsons


def update_compose_labels(compose_cfg: Dict, metadata: Dict[str, str]) -> bool:
    compose_labels = compose_cfg["services"]["osparc-python-runner-tensorflow"]["build"]["labels"]
    changed = False
    for key, value in metadata.items():
        if key in compose_labels:
            if compose_labels[key] == value:
                continue
        compose_labels[key] = value
        changed = True
    return changed


def main(args=None) -> int:
    try:
        parser = argparse.ArgumentParser(description=__doc__)
        parser.add_argument(
            "--compose", help="The compose file where labels shall be updated", type=Path, required=True)
        parser.add_argument("--metadata", help="The metadata yaml file",
                            type=Path, required=False, default="metadata/metadata.yml")
        options = parser.parse_args(args)
        log.info("Testing if %s needs updates using labels in %s",
                 options.compose, options.metadata)
        # get available jsons
        compose_cfg = get_compose_file(options.compose)
        metadata = get_metadata_file(options.metadata)
        json_metadata = stringify_metadata(metadata)
        if update_compose_labels(compose_cfg, json_metadata):
            log.info("Updating %s using labels in %s",
                     options.compose, options.metadata)
            # write the file back
            with options.compose.open('w') as fp:
                yaml.safe_dump(compose_cfg, fp, default_flow_style=False)
                log.info("Update completed")
        else:
            log.info("No update necessary")
        return ExitCode.SUCCESS
    except:  # pylint: disable=bare-except
        log.exception("Unexpected error:")
        return ExitCode.FAIL


if __name__ == "__main__":
    sys.exit(main())


validation/input/input_1/input_script_1.py
import tensorflow as tf
import timeit

import numpy

# Test if there is a GPU availble
device_name = tf.test.gpu_device_name()
if device_name != '/device:GPU:0':
    raise SystemError('GPU device not found')

print('Found GPU at: {}'.format(device_name))

# Run benchmark
device_name = tf.test.gpu_device_name()
if device_name != '/device:GPU:0':
  print(
      '\n\nThis error most likely means that this notebook is not '
      'configured to use a GPU.  Change this in Notebook Settings via the '
      'command palette (cmd/ctrl-shift-P) or the Edit menu.\n\n')
  raise SystemError('GPU device not found')

def cpu():
  with tf.device('/cpu:0'):
    random_image_cpu = tf.random.normal((100, 100, 100, 3))
    net_cpu = tf.keras.layers.Conv2D(32, 7)(random_image_cpu)
    return tf.math.reduce_sum(net_cpu)

def gpu():
  with tf.device('/device:GPU:0'):
    random_image_gpu = tf.random.normal((100, 100, 100, 3))
    net_gpu = tf.keras.layers.Conv2D(32, 7)(random_image_gpu)
    return tf.math.reduce_sum(net_gpu)
  
# We run each op once to warm up; see: https://stackoverflow.com/a/45067900
cpu()
gpu()

# Run the op several times.
print('Time (s) to convolve 32x7x7x3 filter over random 100x100x100x3 images '
      '(batch x height x width x channel). Sum of ten runs.')
print('CPU (s):')
cpu_time = timeit.timeit('cpu()', number=10, setup="from __main__ import cpu")
print(cpu_time)
print('GPU (s):')
gpu_time = timeit.timeit('gpu()', number=10, setup="from __main__ import gpu")
print(gpu_time)
print('GPU speedup over CPU: {}x'.format(int(cpu_time/gpu_time)))